hail/hail/src/is/hail/expr/ir/lowering/IRState.scala

```diff
@@ -1,79 +1,71 @@
 package is.hail.expr.ir.lowering
 
-import is.hail.expr.ir.{BaseIR, TableKeyBy, TableKeyByAndAggregate, TableOrderBy}
+import is.hail.backend.ExecuteContext
+import is.hail.expr.ir.{BaseIR, IRTraversal, TableKeyBy, TableKeyByAndAggregate, TableOrderBy}
 import is.hail.expr.ir.defs.{RelationalLet, RelationalRef}
+import is.hail.expr.ir.lowering.Rules._
+import is.hail.utils.toRichPredicate
+
+abstract class IRState(implicit E: sourcecode.Enclosing) {
+  protected def allows: Rule
+
+  final def verify(ctx: ExecuteContext, ir: BaseIR): Unit =
+    ctx.time {
+      IRTraversal.levelOrder(ir).foreach { ir =>
+        if (!allows(ir))
+          throw new RuntimeException(
+            s"lowered state ${this.getClass.getCanonicalName} forbids IR $ir"
+          )
+      }
+    }
 
-trait IRState {
-
-  val rules: Array[Rule]
```

Comment:
A `Rule` is a predicate. The new `RichPredicate` class adds a cleaner way of combinding predicates.